add unicode decoding for client side to resolve non-English string de…

faith0811 opened the pull request:
…code issues
All string object received by zerorpc is ``str`` type. With English string, this works fine and has no issues at all. But with non-English string, a ``unicode`` object is used instead. It will cause some issue when using ``str`` to interact with ``unicode`` object. 
It is possible to decode string to unicode outside the zerorpc lib, but the most convenient and speedy way is tell ``msgpack`` to use ``unicode`` instead of ``str``. This commit gives the possibility to change the way ``msgpack`` works. It will still use ``str`` for default and gives an option for ``unicode``.

bombela commented:
Hello Wang, thank you for the pull request.

This is basically a hack to interpret all byte strings as utf8. This cannot decently work. As a side effect this will also try to decode zerorpc's protocol strings which are intended to be bytes.

Not all hope is lost though:

See https://github.com/msgpack/msgpack/issues/149 for some interesting details about the new utf8 `string` and `bin` type vs the previous `raw` type.

The catch: the new utf8 `string` type maps to the old `raw` type. This means the old msgpack is not compatible with the new msgpack.

Now, its been a while since the distinction was added and the python implementation supports it (see doc at http://pythonhosted.org/msgpack-python/api.html#msgpack.Packer).

I advise you to set `use_bin_type` to `true` on the packer side instead of unpacking everything in utf8. I am willing take the risk of being incompatible and release a new major version of zerorpc.

faith0811 commented:
Hello bombela, thanks for the advice.
I rewrote the commit and it seems works well.
``msgpack2.0`` needs ``msgpack-python>=0.4.0`` which released in Oct 2013.

And I removed the switch, I think it might be unnecessary in this case, because ``msgpack`` always keeps the original type.

faith0811 commented:
Hi bombela,
If I don't specify the encoding, I will get a ``str`` object encoding with ``utf-8``, not a ``unicode`` object. But the string I put before ``pack`` is a ``unicode`` object.
I updated some tests for details.

bombela commented:
I see, the doc is not clear about it. Looks like its doing the right thing now :) Thank you!
